miden-lib/asm/sat/prologue.masm

```diff
@@ -0,0 +1,363 @@
+#! - cn1_ir is the inputs root of consumed note 1.
+#! - cn1_vr is the vault root of consumed note 1.
+#! - cn1_na is the number of assets in consumed note 1.
+#! - cn1_a1 is the first asset of consumed note 1.
+#! - cn1_a2 is the second asset of consumed note 1.
+proc.process_consumed_note
+    # ingest core note data
+    # ---------------------------------------------------------------------------------------------
+
+    # compute address to store note hash
+    exec.compute_note_data_ptr
+
+    # compute address to store raw note data
+    dup push.1 add
+
+    # absorb note data from the advice provider
+    padw padw padw
+    adv_pipe adv_pipe
+
+    # extract nullifier
+    dropw swapw dropw
+
+    # compute address for nullifier
+    # dup.6 = counter / note index
+    dup.6 exec.compute_nullifier_ptr
+
+
+    # todo assert nullifier does not exist in nullifier db
+
+    # store nullifier in memory and drop from stack
+    mem_storew dropw
+
+    # ingest note assets
+    # ---------------------------------------------------------------------------------------------
+
+    # read the number of assets from the advice provider and store in memory
+    adv_push.1 dup dup.2 mem_store
+
+    # round up the number of assets to the next multiple of 2 (simplifies reading of assets)
+    dup push.1 u32checked_and
+
+    # pad to multiple of two if odd
+    if.true
+        push.1 add
+    end
+
+    # initiate counter for assets
+    push.0
+
+    # prepare address and stack for reading assets
+    movup.2 push.1 add padw padw padw
+
+    # ensure that the number of assets is greater then 0
+    dup.14 dup.14 neq
+
+    # loop and absorb assets from the advice provider
+    while.true
+        # absorb assets from advice provider
+        adv_pipe
+
+        # check if we should loop again
+        movup.13 push.2 add dup movdn.14 dup.15 neq
+    end
+
+    # extract digest from hasher rate elements (h_0, ..., h_3)
+    dropw swapw dropw
+
+    # load expected hash from memory
+    swapw drop drop drop dup movdn.5 push.4 add padw movup.4 mem_loadw
+
+    # assert that the computed hash matches the expected hash
+    assert_eqw
+
+    # compute note hash
+    # ---------------------------------------------------------------------------------------------
+
+    # serial number hash - serial_hash = hmerge(serial_number, 0)
+    dup push.1 add padw movup.4 mem_loadw padw hmerge
+
+    # hash serial_hash with script hash - merge_script = hmerge(serial_hash, script_hash)
+    dup.4 push.2 add padw movup.4 mem_loadw hmerge
+
+    # hash merge_script with inputs hash - recipient = hmerge(merge_script, inputs_hash)
+    dup.4 push.3 add padw movup.4 mem_loadw hmerge
+
+    # hash recipient with vault hash - note_hash = hmerge(recipient, vault_hash)
+    dup.4 push.4 add padw movup.4 mem_loadw hmerge
+
+    # store note hash in memory and clear stack
+    dup.4 mem_storew dropw drop
+
+    # todo - assert note hash exists in note db
```

Comment:
minor nit: I'd prefer using capital letters for `TODO` - makes them stand out a bit more.